proposals/0002-rest-spread-operator.md

```diff
@@ -0,0 +1,133 @@
+# Feature name
+
+* Proposal: [HXP-0002](0002-rest-spread-operator.md)
+* Author: [Thomas Fétiveau](https://github.com/zabojad)
+
+
+## Introduction
+
+Add ES6-like `...` spread and rest operator to the Haxe syntax on both [arrays](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_operator) and [objects](https://babeljs.io/docs/plugins/transform-object-rest-spread/).
+
+
+## Motivation
+
+Currently, there is a gap in Haxe for object duplication and overriding:
+
+* A lot of very popular js libs like React or Redux require developers to implement [pure functions](https://en.wikipedia.org/wiki/Pure_function) massively and this operator makes it really easy to do and avoid over-verbosity of simple operations.
+  
+  This:
+  ```
+  function pure(data : MyStruct) {
+
+  	return {...data, overwrittenFieldName: overwrittenFieldValue};
+  }
+  ```
+
+  is much less verbose than this:
+  ```
+  function pure(data : MyStruct) {
+
+  	var ret : MyStruct = {};
+
+  	for (f in Reflect.fields(data)) {
+
+  		Reflect.setField(ret,f,Reflect.field(data,f));
+  	}
+
+  	ret.overwrittenFieldName = overwrittenFieldValue;
+
+  	return ret;
+  }
+  ```
+
+* The standard javascript externs do not expose the [Object.assign API](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign).
+
+  While it's not complicated to write an extern class for this, it feels wrong to add it to the source code of a custom project while it is something that should be available by default.
+
+* This `...` operator is part of the ECMAScript standard which Haxe syntax follows to some point. Getting closer to the ECMAScript syntax would ease the adoption of Haxe by javascript developers. 
+
+* Also, if added to the Haxe syntax, we would also get several other benefits out of it:
+```
+// array decomposition (non existent in Haxe)
+myFunction(...myArray);
+// shorter and more powerful array literals / comprehension
+var myNewArr = [ 1, 4, ...myOldArr, 56 ];
```

Comment:
Not sure how useful it is to be honest. Also this will obviously require looping, but OTOH I'm not against it, since it looks pretty clear what's happening here.